specs/FrameProcesssInfo.md

```diff
@@ -0,0 +1,540 @@
+    if (firstLevelFrameInfo == frameInfo)
+    {
+        result.append(L" | frame type: first level frame");
+        isMainFrameOrFirstLevelframeInfo = true;
+    }
+    if (!isMainFrameOrFirstLevelframeInfo)
+    {
+        result.append(L" | frame type: other child frame");
+    }
+    
+    COREWEBVIEW2_FRAME_KIND frameKind = COREWEBVIEW2_FRAME_KIND_OTHER;
+    frameInfo2->get_FrameKind(&frameKind);
+    result.append(L"\n | frame kind:" + FrameKindToString(frameKind));
+
+    // Append the frame's direct parent frame's ID if it exists.
+    wil::com_ptr<ICoreWebView2FrameInfo> parentFrameInfo;
+    CHECK_FAILURE(frameInfo2->get_ParentFrameInfo(&parentFrameInfo));
+    if (parentFrameInfo)
+    {
+        CHECK_FAILURE(parentFrameInfo->QueryInterface(IID_PPV_ARGS(&frameInfo2)));
+        CHECK_FAILURE(frameInfo2->get_FrameId(&frameId));
+        result.append(L" | parent frame Id:" + std::to_wstring(frameId));
+    }
+
+    wil::unique_cotaskmem_string sourceRaw;
+    CHECK_FAILURE(frameInfo->get_Source(&sourceRaw));
+    result.append(L"\n | frame source:\n\"");
+    result.append(sourceRaw.get());
+    result.append(L"\"");
+}
+
+// Append the frameInfo's ancestor main frame(webview)'s ID and 
+// ancestor first level frame's ID if it exists.
+void ProcessComponent::AppendAncestorFrameInfo(
+    wil::com_ptr<ICoreWebView2FrameInfo> frameInfo, std::wstring& result)
+{
+    if (frameInfo)
+    {
+        return;
+    }
+
+    wil::com_ptr<ICoreWebView2FrameInfo> mainFrameInfo = GetAncestorMainFrameInfo(frameInfo);
+    wil::com_ptr<ICoreWebView2FrameInfo> firstLevelFrameInfo =
+        GetAncestorFirstLevelFrameInfo(frameInfo);
+    wil::com_ptr<ICoreWebView2FrameInfo2> frameInfo2;
+    UINT32 frameId = 0;
+    if (firstLevelFrameInfo)
+    {
+        CHECK_FAILURE(firstLevelFrameInfo->QueryInterface(IID_PPV_ARGS(&frameInfo2)));
+        CHECK_FAILURE(frameInfo2->get_FrameId(&frameId));
+        result.append(L"\n | ancestor first level frame Id:" + std::to_wstring(frameId));
+    }
+    if (mainFrameInfo)
+    {
+        CHECK_FAILURE(mainFrameInfo->QueryInterface(IID_PPV_ARGS(&frameInfo2)));
+        CHECK_FAILURE(frameInfo2->get_FrameId(&frameId));
+        result.append(L"\n | ancestor main frame Id:" + std::to_wstring(frameId));
+    }
+    result.append(L"},\n");
+}
+
+// Get a string for the frame kind enum value.
+std::wstring ProcessComponent::FrameKindToString(const COREWEBVIEW2_FRAME_KIND kind)
+{
+    switch (kind)
+    {                                                                
+    case kindValue:   
+        return L#kindValue;
+
+        KIND_ENTRY(COREWEBVIEW2_FRAME_KIND_MAIN_FRAME);
+        KIND_ENTRY(COREWEBVIEW2_FRAME_KIND_IFRAME);
+        KIND_ENTRY(COREWEBVIEW2_FRAME_KIND_OTHER);
+    }
+
+    return std::to_wstring(static_cast<uint32_t>(kind));
+}
+```
+C#
+```c#
+string AppendFrameInfo(CoreWebView2FrameInfo frameInfo, string type, string mainFrameId, string firstLevelFrameId) {

```

Comment:
The C++ and C# sample code should more directly match. C++ has different methods than the C# code.

Reply:
Refined.